lib/cn/csched_sp3.c

```diff
@@ -587,21 +587,30 @@ sp3_refresh_thresholds(struct sp3 *sp)
     thresh.rspill_runlen_min = clamp_t(uint8_t, thresh.rspill_runlen_min,
                                        SP3_RSPILL_RUNLEN_MIN, thresh.rspill_runlen_max);
 
-    thresh.rspill_sizemb_max = clamp_t(uint16_t, thresh.rspill_sizemb_max,
-                                       SP3_RSPILL_SIZEMB_MIN, SP3_RSPILL_SIZEMB_MAX);
+    thresh.rspill_wlen_max = clamp_t(size_t, thresh.rspill_wlen_max,
+                                     SP3_RSPILL_WLEN_MIN, SP3_RSPILL_WLEN_MAX);
 
     /* leaf node compaction settings */
     v = sp->rp->csched_leaf_comp_params;
     if (v) {
         thresh.lcomp_runlen_max = (v >> 0) & 0xff;
         thresh.lcomp_split_pct = (v >> 16) & 0xff;
-        thresh.lcomp_split_keys = (v >> 24) & 0xff;
+        thresh.lcomp_split_keys = ((v >> 24) & 0xff) << 22;
     } else {
-        thresh.lcomp_runlen_max = SP3_LCOMP_RUNLEN_MAX;
-        thresh.lcomp_split_pct = SP3_LCOMP_SPLIT_PCT;
-        thresh.lcomp_split_keys = SP3_LCOMP_SPLIT_KEYS; /* units of 4 million */
+        thresh.lcomp_runlen_max = SP3_LCOMP_RUNLEN_MAX_DEFAULT;
+        thresh.lcomp_split_pct = SP3_LCOMP_SPLIT_PCT_DEFAULT;
+        thresh.lcomp_split_keys = SP3_LCOMP_SPLIT_KEYS_DEFAULT;
     }
 
+    thresh.lcomp_runlen_max = clamp_t(uint, thresh.lcomp_runlen_max,
+                                      SP3_LCOMP_RUNLEN_MAX_MIN, SP3_LCOMP_RUNLEN_MAX_MAX);
+
+    thresh.lcomp_split_pct = clamp_t(uint, thresh.lcomp_split_pct,
+                                     SP3_LCOMP_SPLIT_PCT_MIN, SP3_LCOMP_SPLIT_PCT_MAX);
+
+    thresh.lcomp_split_keys = clamp_t(uint, thresh.lcomp_split_keys,
+                                      SP3_LCOMP_SPLIT_KEYS_MIN, SP3_LCOMP_SPLIT_KEYS_MAX);
```

Comment:
These thresholds are for developer use only and should not be clamped.  Developers may want to run with thresholds set to 0 or UINT_MAX to disable a certain behavior.

Reply:
As an example, I wanted to run a test with a queue disabled, I used to be able to set thread count to 0, but now it is clamped to 1,16. 
```
static inline uint
qthreads(struct sp3 *sp, uint qnum)
{
    uint64_t rparam = sp->rp->csched_qthreads;
    uint n;

    n = (rparam >> (qnum * 8)) & 0xff;

    return clamp_t(uint, n, 1, 16);
}
```

Reply:
Nothing prevents users from changing them, and these rparams are so difficult to use it's hard for even a dev to get right.
But it if makes you happy I'll just set the limits to UNIT_MAX for now then and leave the clamping code.  It might be better to leave things as they are here, but use an rparam to disable the clamping (i.e., users who use that get no help from us!).

Reply:
I was just observing that clamping reduces flexibility for developers.  I never imagined these params being used by anyone outside our team.

Reply:
Yes, agreed.  I'm changing the limits to allow that, testing first!

Reply:
That's 16 threads per queue, so 96 total compaction threads in various forms...

Reply:
Alex, please review most recent commit.